sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/jmh/Main.java

```diff
@@ -0,0 +1,150 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.testutils.jmh;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+import static org.openjdk.jmh.annotations.Mode.SingleShotTime;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.testutils.publishing.InfluxDBPublisher;
+import org.apache.beam.sdk.testutils.publishing.InfluxDBPublisher.DataPoint;
+import org.apache.beam.sdk.testutils.publishing.InfluxDBSettings;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.infra.BenchmarkParams;
+import org.openjdk.jmh.results.BenchmarkResult;
+import org.openjdk.jmh.results.BenchmarkResultMetaData;
+import org.openjdk.jmh.results.Result;
+import org.openjdk.jmh.results.RunResult;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.CommandLineOptionException;
+import org.openjdk.jmh.runner.options.CommandLineOptions;
+
+/**
+ * Custom main wrapper around the {@link Runner JMH runner} that supports publishing benchmarks to
+ * InfluxDB.
+ *
+ * <p>If {@link InfluxDBSettings} can be inferred from the environment, benchmark results will be
+ * published to InfluxDB. Otherwise this will just delegate to the default {@link
+ * org.openjdk.jmh.Main JMH Main} class.
+ *
+ * <p>Use the following environment variables to configure {@link InfluxDBSettings}:
+ *
+ * <ul>
+ *   <li>{@link #INFLUXDB_HOST}
+ *   <li>{@link #INFLUXDB_DATABASE}
+ *   <li>{@link #INFLUXDB_MEASUREMENT}
+ * </ul>
+ */
+public class Main {
+  private static final String INFLUXDB_HOST = "INFLUXDB_HOST";
+  private static final String INFLUXDB_DATABASE = "INFLUXDB_DATABASE";
+  private static final String INFLUXDB_MEASUREMENT = "INFLUXDB_MEASUREMENT";
+
+  public static void main(String[] args)
+      throws CommandLineOptionException, IOException, RunnerException {
+    final CommandLineOptions opts = new CommandLineOptions(args);
+    final InfluxDBSettings influxDB = influxDBSettings();
+
+    if (influxDB == null
+        || isSingleShotTimeOnly(opts.getBenchModes())
+        || opts.shouldHelp()
+        || opts.shouldList()
+        || opts.shouldListWithParams()
+        || opts.shouldListProfilers()
+        || opts.shouldListResultFormats()) {
+      // delegate to JMH runner
+      org.openjdk.jmh.Main.main(args);
+      return;
+    }
+
+    final Runner runner = new Runner(opts);
+    final Collection<RunResult> results = runner.run();
+
+    final Collection<DataPoint> dataPoints =
+        results.stream()
+            .filter(r -> r.getParams().getMode() != SingleShotTime)
```

Comment:
JMH would fail hard (exceptionally) in that case, so no results are available and nothing gets published.

Reply:
Is that because we currently configure the `-foe` flag (fail on error)?

Reply:
yes, that's true... 
if `-foe` is disabled `Collection<RunResult> results` will only contain results with successful forks. in case all forks for a benchmark fail it won't be included in results.

it probably makes sense to add `sampleCount` as an additional field to capture this in case `-foe` is disabled.

Reply:
Looks good to me to record the sample count based upon the code that you have added instead of requiring exactly X samples. We can always update dashboards to filter on partial successes if necessary.